Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -6483,30 +6489,433 @@ bool Parser::has_ignored_vendor_prefix(StringView string)
+            Vector<ComponentValue> function_values;
+            TokenStream source_function_contents { source_function.values() };
+            if (!expand_unresolved_values(element, property_name, source_function_contents, function_values))
+                return false;
+            NonnullRefPtr<Function> function = Function::create(source_function.name(), move(function_values));
+            dest.empend(function);
+            continue;
+        }
+        if (value.is_block()) {
+            auto const& source_block = value.block();
+            TokenStream source_block_values { source_block.values() };
+            Vector<ComponentValue> block_values;
+            if (!expand_unresolved_values(element, property_name, source_block_values, block_values))
+                return false;
+            NonnullRefPtr<Block> block = Block::create(source_block.token(), move(block_values));
+            dest.empend(move(block));
+            continue;
+        }
+        dest.empend(value.token());
+    }
+
+    return true;
+}
+
+// https://drafts.csswg.org/css-values-5/#attr-substitution
+bool Parser::substitute_attr_function(DOM::Element& element, StringView property_name, Function const& attr_function, Vector<ComponentValue>& dest)
+{
+    // First, parse the arguments to attr():
+    // attr() = attr( <q-name> <attr-type>? , <declaration-value>?)
+    // <attr-type> = string | url | ident | color | number | percentage | length | angle | time | frequency | flex | <dimension-unit>
+    TokenStream attr_contents { attr_function.values() };
+    attr_contents.skip_whitespace();
+    if (!attr_contents.has_next_token())
+        return false;
+
+    // - Attribute name
+    // FIXME: Support optional attribute namespace
+    if (!attr_contents.peek_token().is(Token::Type::Ident))
+        return false;
+    auto attribute_name = attr_contents.next_token().token().ident();
+    attr_contents.skip_whitespace();
+
+    // - Attribute type (optional)
+    auto attribute_type = "string"_fly_string;
+    if (attr_contents.peek_token().is(Token::Type::Ident)) {
+        attribute_type = MUST(FlyString::from_utf8(attr_contents.next_token().token().ident()));
+        attr_contents.skip_whitespace();
+    }
+
+    // - Comma, then fallback values (optional)
+    bool has_fallback_values = false;
+    if (attr_contents.has_next_token()) {
+        if (!attr_contents.peek_token().is(Token::Type::Comma))
+            return false;
+        (void)attr_contents.next_token(); // Comma
+        has_fallback_values = true;
+    }
+
+    // Then, run the substitution algorithm:
+
+    // 1. If the attr() function has a substitution value, replace the attr() function by the substitution value.
+    // https://drafts.csswg.org/css-values-5/#attr-types
+    if (element.has_attribute(attribute_name)) {
+        auto attribute_value = element.get_attribute_value(attribute_name);
+        if (attribute_type.equals_ignoring_ascii_case("angle"_fly_string)) {
```

Comment:
Should this be:
```suggestion
        if (attribute_type.equals_ignoring_ascii_case("angle"sv)) {
```
Since this takes a StringView?

Reply:
There's a `FlyString` overload for that function too. I was thinking that comparing the two FlyStrings would be faster in the usual case (both lowercase) but actually it probably doesn't help. :thonk:

Reply:
ah right! i missed thay overload 😅